lighter/callbacks/writer/base.py

```diff
@@ -1,146 +1,179 @@
+from lighter.callbacks.utils import flatten_structure
 
 
 class LighterBaseWriter(ABC, Callback):
-    """Base class for a Writer. Override `self.write()` to define how a prediction should be saved.
-    `LighterBaseWriter` sets up the write directory, and defines `on_predict_batch_end` and
-    `on_predict_epoch_end`. `write_interval` specifies which of the two should the writer call.
+    """
+    Base class for defining custom Writer. It provides the structure to save predictions in various formats.
+
+    Subclasses should implement:
+        1) `self._writers` attribute to specify the supported formats and their corresponding writer functions.
+        2) `self.write()` method to specify the saving strategy for a prediction.
 
     Args:
-        write_dir (str): the Writer will create a directory inside of `write_dir` with date
-            and time as its name and store the predictions there.
-        write_format (Optional[Union[str, List[str], Dict[str, str], Dict[str, List[str]]]]):
-            type in which the predictions will be stored. Passed automatically to the `write()`
-            abstract method and can be used to support writing different types. Should the Writer
-            support only one type, this argument can be removed from the overriden `__init__()`'s
-            arguments and set `self.write_format = None`.
-        write_interval (str, optional): whether to write on each step or at the end of the prediction epoch.
-            Defaults to "step".
+        directory (str): Base directory for saving. A new sub-directory with current date and time will be created inside.
+        format (Optional[Union[str, List[str], Dict[str, str], Dict[str, List[str]]]]):
+            Desired format(s) for saving predictions. The format will be passed to the `write` method.
+        additional_writers (Optional[Dict[str, Callable]]): Additional writer functions to be registered with the base writer.
     """
 
     def __init__(
         self,
-        write_dir: str,
-        write_format: Optional[Union[str, List[str], Dict[str, str], Dict[str, List[str]]]],
-        write_interval: str = "step",
+        directory: str,
+        format: Optional[Union[str, List[str], Dict[str, str], Dict[str, List[str]]]],
+        additional_writers: Optional[Dict[str, Callable]] = None,
     ) -> None:
-        self.write_dir = Path(write_dir) / datetime.now().strftime("%Y%m%d_%H%M%S")
-        self.write_format = write_format
-        self.write_interval = write_interval
+        # Create a unique directory using the current date and time
+        self.directory = Path(directory) / datetime.now().strftime("%Y%m%d_%H%M%S")
+        self.format = format
+
+        # When IDs are not provided, keep track of the global prediction count. Supports DDP.
+        self._pred_count = None
 
-        self.parsed_write_format = None
+        # Ensure that default writers are defined
+        if not hasattr(self, "_writers"):
```

Comment:
could you try implementing it? I tried and didn't come up with a good solution. You can experiment with the config in the cifar10 project.

When overriding, you cannot define a property that will just return the writers dict because you cannot update the dict inside of it then. you would have to do something like this:
```python
class MyClass:
    def __init__(self):
        self._my_dict = {"a": 1, "b": 2}

    @property
    def my_dict(self):
        return self._my_dict
```
which actually doesn't do any good in our case

Reply:
Let me try out some stuff!